### Extract data of Toronto neighborhoods from Wikipedia, clean and display the top 10 rows along with shape head

#### Import libraries by use pandas and the BeautifulSoup package, (Note that , you can use any  comfortable with to transform the data in the table on the Wikipedia page into the above pandas dataframe.)

In [1]:
from bs4 import BeautifulSoup
import requests
import pandas as pd

In [2]:
# download url data from internet
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(url).text
Canada_data = BeautifulSoup(source, 'lxml')

In [3]:
# creat a new Dataframe
column_names = ['Postalcode','Borough','Neighborhood']
toronto = pd.DataFrame(columns = column_names)

In [4]:
content = Canada_data.find('div', class_='mw-parser-output')
table = content.table.tbody
postcode = 0
borough = 0
neighborhood = 0

for tr in table.find_all('tr'):
    i = 0
    for td in tr.find_all('td'):
        if i == 0:
            postcode = td.text
            i = i + 1
        elif i == 1:
            borough = td.text
            i = i + 1
        elif i == 2: 
            neighborhood = td.text.strip('\n').replace(']','')
    toronto = toronto.append({'Postalcode': postcode,
                              'Borough': borough,
                              'Neighborhood': neighborhood},
                               ignore_index=True )

# clean dataframe 
toronto = toronto[toronto.Borough!='Not assigned\n']
toronto = toronto[toronto.Borough!= 0]
toronto.reset_index(drop = True, inplace = True)
i = 0
for i in range(0,toronto.shape[0]):
    if toronto.iloc[i][2] == 'Not assigned\n':
        toronto.iloc[i][2] = toronto.iloc[i][1]
        i = i+1

df = toronto.groupby(['Postalcode','Borough'])['Neighborhood'].apply(', '.join).reset_index()
df.head()

Postalcode        Borough                              Neighborhood
0      M1B\n  Scarborough\n                           Malvern / Rouge
1      M1C\n  Scarborough\n  Rouge Hill / Port Union / Highland Creek
2      M1E\n  Scarborough\n       Guildwood / Morningside / West Hill
3      M1G\n  Scarborough\n                                    Woburn
4      M1H\n  Scarborough\n                                 Cedarbrae

### Next Let's remove '\n' in the Borough Column 

In [5]:
df = df.replace('\n','', regex=True)

In [6]:
df.head(10)

Postalcode      Borough                                       Neighborhood
0        M1B  Scarborough                                    Malvern / Rouge
1        M1C  Scarborough           Rouge Hill / Port Union / Highland Creek
2        M1E  Scarborough                Guildwood / Morningside / West Hill
3        M1G  Scarborough                                             Woburn
4        M1H  Scarborough                                          Cedarbrae
5        M1J  Scarborough                                Scarborough Village
6        M1K  Scarborough      Kennedy Park / Ionview / East Birchmount Park
7        M1L  Scarborough                  Golden Mile / Clairlea / Oakridge
8        M1M  Scarborough  Cliffside / Cliffcrest / Scarborough Village West
9        M1N  Scarborough                       Birch Cliff / Cliffside West

In [7]:
def neighborhood_list(grouped):    
    return ', '.join(sorted(grouped['Neighborhood'].tolist()))
                    
grp = df.groupby(['Postalcode', 'Borough'])
df2 = grp.apply(neighborhood_list).reset_index(name='Neighborhood')

In [8]:
df2.head()

Postalcode      Borough                              Neighborhood
0        M1B  Scarborough                           Malvern / Rouge
1        M1C  Scarborough  Rouge Hill / Port Union / Highland Creek
2        M1E  Scarborough       Guildwood / Morningside / West Hill
3        M1G  Scarborough                                    Woburn
4        M1H  Scarborough                                 Cedarbrae

In [9]:
print(df2.shape)
df2.head(10)

(103, 3)


Postalcode      Borough                                       Neighborhood
0        M1B  Scarborough                                    Malvern / Rouge
1        M1C  Scarborough           Rouge Hill / Port Union / Highland Creek
2        M1E  Scarborough                Guildwood / Morningside / West Hill
3        M1G  Scarborough                                             Woburn
4        M1H  Scarborough                                          Cedarbrae
5        M1J  Scarborough                                Scarborough Village
6        M1K  Scarborough      Kennedy Park / Ionview / East Birchmount Park
7        M1L  Scarborough                  Golden Mile / Clairlea / Oakridge
8        M1M  Scarborough  Cliffside / Cliffcrest / Scarborough Village West
9        M1N  Scarborough                       Birch Cliff / Cliffside West

In [10]:
print('The DataFrame shape is', df2.shape)

The DataFrame shape is (103, 3)


#### The dataframe has 103 Postal codes but it has 212 rows, because each Postal code can present more than one neighborhood (210 in total). Therefore, the dataframe should be group by the Postal code, ending with a dataframe with 103 rows.